packages/sdk/src/tests/multi_ne/trading.test.ts

```diff
@@ -0,0 +1,182 @@
+        const transferEvent: PlainMessage<BlockchainTransaction_Transfer> = {
+            amount: amountToTransfer.toString(),
+            address: bin_fromHexString(tokenAddress),
+            sender: bin_fromHexString(charlieClient.userId),
+            messageId: bin_fromHexString(threadParentId),
+            channelId: bin_fromHexString(channelId),
+        }
+
+        await expect(
+            charlieClient.addTransaction_Transfer(31337, receipt, transferEvent),
+        ).rejects.toThrow('matching transfer event not found in receipt logs')
+    })
+
+    test('should accept token transfers where the user == from', async () => {
+        // this is a transfer event from bob, he's the sender (from)
+        const transferEvent: PlainMessage<BlockchainTransaction_Transfer> = {
+            amount: amountToTransfer.toString(),
+            address: bin_fromHexString(tokenAddress),
+            sender: bin_fromHexString(bobsClient.userId),
+            messageId: bin_fromHexString(threadParentId),
+            channelId: bin_fromHexString(channelId),
+        }
+
+        const { eventId } = await bobsClient.addTransaction_Transfer(31337, receipt, transferEvent)
+        expect(eventId).toBeDefined()
+
+        await waitFor(() =>
+            expect(
+                bobsClient.streams.get(channelId)?.view.timeline.some((m) => m.hashStr === eventId),
+            ).toBeDefined(),
+        )
+    })
+
+    test('should accept token transfers where the user == to', async () => {
+        // this is a transfer event to alice, she's the recipient (to)
+        const transferEvent: PlainMessage<BlockchainTransaction_Transfer> = {
+            amount: amountToTransfer.toString(),
+            address: bin_fromHexString(tokenAddress),
+            sender: bin_fromHexString(aliceClient.userId),
+            messageId: bin_fromHexString(threadParentId),
+            channelId: bin_fromHexString(channelId),
+        }
+
+        const { eventId } = await aliceClient.addTransaction_Transfer(31337, receipt, transferEvent)
+        expect(eventId).toBeDefined()
+
+        await waitFor(() =>
+            expect(
+                bobsClient.streams.get(channelId)?.view.timeline.some((m) => m.hashStr === eventId),
+            ).toBeDefined(),
+        )
+    })
+
+    test('should reject duplicate transfers', async () => {
+        // alice can't add the same transfer event twice
+        const transferEvent: PlainMessage<BlockchainTransaction_Transfer> = {
+            amount: amountToTransfer.toString(),
+            address: bin_fromHexString(tokenAddress),
+            sender: bin_fromHexString(aliceClient.userId),
+            messageId: bin_fromHexString(threadParentId),
+            channelId: bin_fromHexString(channelId),
+        }
+
+        await expect(
+            aliceClient.addTransaction_Transfer(31337, receipt, transferEvent),
+        ).rejects.toThrow('duplicate transaction')
+    })
```

Comment:
you can add a test - "transaction shows up in channel stream" I've got one in tips. It has a goofy waitfor, but it's just checking the stream events.